`Find All references` doesn't exist for typescript plugin #14856

vrubezhny opened the pull request:
Signed-off-by: Victor Rubezhny <vrubezhny@redhat.com>

<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->
The fix adds the required API and fixes required to make possible to run VSCode Reference View extension plug-in on Theia.

![et14856](https://user-images.githubusercontent.com/620781/73654882-0c5d5e00-468d-11ea-82c5-fad6a1503d00.gif)

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
Tested manually using references-view-0.0.47 (https://github.com/microsoft/vscode-references-view) downloaded into `plugins/` Theia folder, 

#### Review checklist

- [*] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)
- Check that VSCode Reference View is loaded and shown in Theia Plugins View
- Test by executing `Find All References` context menu action and revisiting the result tree of the References View
- Test by executing `Find All Implementations` context menu action and revisiting the result tree of the References View

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

akosyakov commented on packages/plugin-ext/src/plugin/types-impl.ts:
```diff
@@ -116,6 +116,20 @@ export namespace TextEditorSelectionChangeKind {
 }
 
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function es5ClassCompat(target: Function): any {
```
we have it already somewhere

vrubezhny replied:
yes. I can use it from:
```
import { es5ClassCompat } from '../common/types';
```

akosyakov commented:
I can have a look but later this week. That's cool btw :)

Tree styling on the screenshot looks bad.

tsmaeder commented on packages/plugin-ext/src/plugin/types-impl.ts:
```diff
@@ -271,17 +286,23 @@ export class Position {
+    }
 }
 
+@es5ClassCompat
```
Why do we need this? Was the code broken before?

vrubezhny replied:
Probably not. it wasn't broken. But (while I was having problems to build theia) I decided to make these classes to exact copies of according VS Code classes. The main changes here were absence of these `@es5ClassCompat` annotations and the absence of `toJSON(): any` methods (and some other methods).
But in theory I can revert these changes back. (Not sure if it would work, but it's easy to test)

tsmaeder replied:
@akosyakov can you shed some light on this?

akosyakov replied:
It would be required if we down compile to es6, but code is still executed in es5 envs, then it is patching es6 classes to look like es5 classes. It is not interesting for us right now, since we still down compile to es5, but should not harm to copy.

vrubezhny replied:
I have removed this change in the last version of the PR, so no changes in `types-impl.ts`

tsmaeder commented on packages/plugin-ext/src/plugin/types-impl.ts:
```diff
@@ -390,7 +411,8 @@ export class Range {
 
-    static isRange(thing: {}): thing is theia.Range {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    static isRange(thing: any): thing is theia.Range {
```
Why is this change necessary?

vrubezhny replied:
The same reason: to make this class to be defined exactly as it's done in VS Code. Probably it's not necessary change.

vrubezhny replied:
I can rollback the changes in types-impl.ts as they aren't necessary and do not bring any fixes/improvements regarding this issue. Rolledback in fixup

tsmaeder commented:
I built your branch and added the references extension from the marketplace. I noticed the following problems:
![image](https://user-images.githubusercontent.com/13163770/73941182-0bc30280-48ed-11ea-8306-ad957f09e035.png)
1. The View has no icon in the view sidebar
2. the search term ("CallHierarchyCommands") is  not highlighted
3. The icon is not the typescript file icon, but a "folder" icon.
4. There is a line break in the rendereing of the file entry ("callhierarchy-contribution.ts\npackages...")
5. The message "4 results in 1 file is not indented, which looks weird.
6. When I do "Show All References" from the context menu in an editor, the references view does not open, I need to go through the menu bar "view" menu.

vince-fugnitto commented:
> 1. The View has no icon in the view sidebar

For that specific issue, there is a bug tracking **icons** which are not rendered in Electron: https://github.com/eclipse-theia/theia/issues/7040

vrubezhny commented on packages/plugin-ext-vscode/src/browser/plugin-vscode-commands-contribution.ts:
```diff
@@ -374,8 +375,36 @@ export class PluginVscodeCommandsContribution implements CommandContribution {
+            {
+                execute: ((resource: URI, position: Position) => {
+                    const args = {
+                        resource: monaco.Uri.parse(resource.toString()),
```
Probably, but how can I make `monaco.URI` imported/available? Bexause:
```
packages/plugin-ext-vscode/src/browser/plugin-vscode-commands-contribution.ts:388:42 - 
error TS2551: Property 'URI' does not exist on type 'typeof monaco'. Did you mean 'Uri'?

388                         resource: monaco.URI.from(resource),
```

tsmaeder replied:
I mean monaco.Uri

vrubezhny replied:
OK

vrubezhny commented on packages/plugin-ext/src/common/index.ts:
```diff
@@ -18,4 +18,4 @@
 export * from './plugin-api-rpc';
 export * from './plugin-ext-api-contribution';
-export * from './known-commands';
+export * from '../plugin/known-commands';
```
OK

vrubezhny commented on packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts:
```diff
@@ -268,6 +268,20 @@ export class PluginViewRegistry implements FrontendApplicationContribution {
+        const focusCommandId = `${view.id}.focus`;
+        toDispose.push(this.commands.registerCommand({
+            id: focusCommandId,
+            label: 'Focus ' + view.name + ' View'
```
Not intensional... Just followed the pattern for `toggleCommandId` in lines 208-211 of the same file.

vrubezhny replied:
But actually specifying a property, like `label: 'Focus ${view.name} View'` doesn't work. It gives:
![Screenshot from 2020-02-07 16-13-51](https://user-images.githubusercontent.com/620781/74041137-63cd3800-49c5-11ea-85df-3f7f4cdc0921.png)
instead of:
![Screenshot from 2020-02-07 16-16-43](https://user-images.githubusercontent.com/620781/74041203-82cbca00-49c5-11ea-8a0f-dcfd8d63e972.png)
So, probably the only way to make a label here is the concatenation.

benoitf replied:
did you use backtick ` ?

vrubezhny replied:
Yes. I can make it to be like focusCommandId...

vrubezhny replied:
> did you use backtick ` ?
No, I didn't. That's why it didn't work for me. Now it's fixed (I'm preparing the fixup)

vrubezhny commented on packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts:
```diff
@@ -268,6 +268,20 @@ export class PluginViewRegistry implements FrontendApplicationContribution {
             }
         }));
+
+        const focusCommandId = `${view.id}.focus`;
```
And yes, I can ommit the constant and make it to be inline.

vrubezhny commented on packages/plugin-ext/src/main/browser/view/tree-views-main.ts:
```diff
@@ -108,6 +108,23 @@ export class TreeViewsMainImpl implements TreeViewsMain, Disposable {
     }
 
+    async $setMessage(treeViewId: string, message: string): Promise<void> {
+        this.viewRegistry.getView(treeViewId).then(viewPanel => {
```
Not sure if we can do anything here (with `$setMessage` or `$setTitle`) in case of tree view doesn't exist. Add an error record to Log?

tsmaeder replied:
Yes, this indicates something is wrong, needs to got to the log even if we can't handle it.

vrubezhny replied:
When I try to print an error by `console.error(...)` TSLint says: `Calls to 'console.error' are not allowed. (no-console)tslint(1)`. 
What is the correct way o report an error here?

vrubezhny commented on packages/plugin-ext/src/main/browser/view/tree-views-main.ts:
```diff
@@ -108,6 +108,23 @@ export class TreeViewsMainImpl implements TreeViewsMain, Disposable {
+        this.viewRegistry.getView(treeViewId).then(viewPanel => {
+            const widget = viewPanel && viewPanel.widgets[0];
+            if (widget instanceof TreeViewWidget) {
+                widget.message = message;
```
Yes I've followed the patterns of existing methods. And actually I don't see any valuable error processing here other than just ignoring in case of absent tree view.

vrubezhny commented on packages/plugin-ext/src/plugin/known-commands.ts:
```diff
@@ -56,207 +57,226 @@ export namespace KnownCommands {
-    mappings['goToNextReference'] = ['goToNextReference', MONACO_CONVERSION_IDENTITY];
-    mappings['goToPreviousReference'] = ['goToPreviousReference', MONACO_CONVERSION_IDENTITY];
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const VSCODE_CONVERSION_IDENTITY = (args: any[] | undefined) => {
```
I've got rid of that necessity to use `undefined` as second argument in case of result conversion is not needed. So readability of git diff is improved

vrubezhny commented on packages/plugin-ext/src/plugin/known-commands.ts:
```diff
@@ -317,6 +337,18 @@ export namespace KnownCommands {
     }
 
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/tslint/config
+    function vscodeArgsConverter(args: any[]) {
```
nope. It converts the result of `execute...`-like command to theia types (a la vscode types)

tsmaeder replied:
What kind of types are in those results?

vrubezhny replied:
Here the results are to be converted from `model.Location` to `types.Location` (which should be equal to `vscode.Location`)

vrubezhny commented on packages/plugin-ext/src/plugin/type-converters.ts:
```diff
@@ -585,6 +586,36 @@ export function toDocumentSymbol(symbol: model.DocumentSymbol): theia.DocumentSy
+    if (!thing) {
+        return false;
+    }
+    return ('startLineNumber' in thing) && ('startColumn' in thing) &&
```
How? I use these methods (isModelLocation(), isModelRange() and isUriComponents()) in result conversion in order to detect if I can convert an object into a theia.Location, theia.Range etc.  Here I have a json-like object with properties. No type nor type name available...

tsmaeder replied:
still, doesn't `startLineNumber` have to be of type `number`?

vrubezhny replied:
added checks for property types

akosyakov commented:
> There is a line break in the rendereing of the file entry ("callhierarchy-contribution.ts\npackages...")

It is a regression: https://github.com/eclipse-theia/theia/issues/7109

vrubezhny commented on packages/plugin-ext/src/plugin/known-commands.ts:
```diff
@@ -56,207 +57,226 @@ export namespace KnownCommands {
-    mappings['goToNextReference'] = ['goToNextReference', MONACO_CONVERSION_IDENTITY];
-    mappings['goToPreviousReference'] = ['goToPreviousReference', MONACO_CONVERSION_IDENTITY];
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const VSCODE_CONVERSION_IDENTITY = (args: any[] | undefined) => {
```
Renamed

vrubezhny commented:
I've uploaded a new version of PR. Most of requests to change were reworked, but not all. For example, https://github.com/eclipse-theia/theia/pull/7055#discussion_r375772277 - to improve KnownCommands.map()...
Also, I found that I still have problem to show a search message on very first run of 'Find References/Implementations' - the message isn't shown... Looking into why this happens.  
But still, please, review

tsmaeder commented on packages/plugin-ext/src/plugin/known-commands.ts:
```diff
@@ -56,207 +57,226 @@ export namespace KnownCommands {
+        }
+    };
+
+    mappings['editor.action.select.all'] = ['editor.action.select.all', MONACO_CONVERSION_IDENTITY, undefined];
```
But these are refactorings of existing code that are not really related to the task at hand. We could address that at a later time.

vrubezhny commented:
I've pushed a new version of PR:
- Fixed `very first command invocation` search message problem: not the message is displayed on the first run
- Added check for type when converting from ` model.* ` to ` types.*` for `Location`, `Range` and `UriComponents` 
- An error is reported to console in `$setTitle` and `$setMessage` API in case of Tree Widget cannot be found

@tsmaeder Please review.

akosyakov commented on packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts:
```diff
@@ -268,6 +268,19 @@ export class PluginViewRegistry implements FrontendApplicationContribution {
             }
         }));
+
+        toDispose.push(this.commands.registerCommand({
```
Is it something to align with VS Code? Otherwise we already have commands to toggle and open views. I don't think we need more.

vrubezhny replied:
Exactly. when a View is Registered in VSCode a `Focus` command for such a view is registered automatically. This is a standard VSCode behaviour. 
In case of VSCode Reference View` plugin this `Focus` command is used to open/activate the References View on execution of `Find All References` and `Find All Implementations` commands. Without having  this command added an error is throws as well as the View is not activated. In the plugin it is used here:
https://github.com/microsoft/vscode-references-view/blob/master/src/extension.ts#L23

akosyakov commented on packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx:
```diff
@@ -372,4 +372,24 @@ export class TreeViewWidget extends TreeWidget {
+
+    protected renderTreeInfo(): React.ReactNode {
+        if (this._message) {
+            return <div className='search-info'>{this._message}</div>;
```
we should not rely on classes of completely other components

vrubezhny replied:
Can anybody help with searching/selection of suitable class here?

akosyakov replied:
You should be able to define a new css class?

akosyakov commented on packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx:
```diff
@@ -372,4 +372,24 @@ export class TreeViewWidget extends TreeWidget {
+        return React.createElement('div', this.createContainerAttributes(), this.renderTreeInfo(), this.renderTree(this.model));
+    }
+
+    private _message: string | undefined;
```
Why cannot be message be represented as first tree node with message as caption? then it will automatically get proper styling

vrubezhny replied:
Can you point me to an example of this, if any. Maybe you right... I didn't tried to do this yet, but if it is possible to make such a node not to be selectable and disable any actions on it (like expand/open resource and so on), it may be a better solution.

tsmaeder replied:
Well, the styling seems different in VS Code for the message, so if we want to align with that behaviour we need to keep it separate. Also, I guess we'd have to filter events to to that first item so the don't get passed to whoever might listen to them. I don't think it's saving work, in the end.

akosyakov replied:
ok

akosyakov commented on packages/plugin-ext/src/main/browser/view/tree-views-main.ts:
```diff
@@ -108,6 +108,28 @@ export class TreeViewsMainImpl implements TreeViewsMain, Disposable {
+
+        if (widget instanceof TreeViewWidget) {
+            widget.message = message;
+        } else {
```
i'm not sure about logging errors in else cases, other methods don't do it. It could be that message is resolved by the time when a user already closed the tree, there is nothing bogus to ignore message then.

vrubezhny replied:
It was requested by @tsmaeder. I also think the logging is not needed here. 
During the run this error appears several times, while not always indicate a real problem with absent widget

tsmaeder replied:
So what we're trying to guard here is the widget being present, not that it might be there but not a TreeWidget? Then why not do a `if (widget)` It has the same effect, but IMO would be better express the meaning.

akosyakov replied:
If we like here https://github.com/eclipse-theia/theia/pull/7055#discussion_r381852705 then we should log error if only metadata is missing. The same for title.

akosyakov commented on packages/plugin-ext/src/main/browser/view/tree-views-main.ts:
```diff
@@ -108,6 +108,28 @@ export class TreeViewsMainImpl implements TreeViewsMain, Disposable {
+
+        if (widget instanceof TreeViewWidget) {
+            widget.message = title;
+        } else {
```
the same as for message

vrubezhny replied:
Agreed.

Actually there is an error here. Sorry, the title is to be set differently. Fixed

akosyakov commented on packages/plugin-ext/src/plugin/command-registry.ts:
```diff
@@ -104,8 +104,20 @@ export class CommandRegistryImpl implements CommandRegistryExt {
+            return KnownCommands.map(id, args, (mappedId: string, mappedArgs: any[] | undefined, mappedResult: KnownCommands.ConversionFunction) => {
+                const mr: KnownCommands.ConversionFunction = mappedResult;
+                return this.proxy.$executeCommand(mappedId, ...mappedArgs).then((result: any) => {
+                    if (!Array.isArray(result) || result === undefined) {
```
Could someone elaborate what everything what is not array should be ignored? I cannot follow why this code has to be changed. It's used extensively from many places and custom commands obviously can return something what is not array.

vrubezhny replied:
Command arguments and command result currently are treated as arrays of objects. If array element is elected to be converted, according conversion function will be used to convert the type of such element. If there is no conversion function defined for an element type - such element is to stay unchanged.
Command arguments are always an array,.
Two `vscode.execute...` commands that were added always return an array of Location objects.
So, currently I don't know any commands returning not an array as result and requiring the result conversion. 
Probably this may be changed in future.

tsmaeder replied:
Command results should not be treated as arrays: whether a command returns an array or a single object or even just a number is up to the command. We can construct composite conversion functions for array results in `KnownCommands`. This code here should not know anything about concrete commands.

akosyakov replied:
It seems to be hold true that monaco commands are always returning undefined (void) and exposed vscode commands are always return array or undefined. Although a type says that it can be anything: https://github.com/microsoft/vscode/blob/355eea22ea9a98a5d0517ffeae5bd176e96bb185/src/vs/workbench/api/common/extHostApiCommands.ts#L45

I'm concerned because code is not easy to grasp here. Anyway if you sure that it does not break let's go with it.

akosyakov commented on packages/plugin-ext/src/plugin/known-commands.ts:
```diff
@@ -17,9 +17,9 @@
-import { cloneAndChange } from './objects';
-import { Position, Range, Location } from '../plugin/types-impl';
-import { fromPosition, fromRange, fromLocation } from '../plugin/type-converters';
+import { cloneAndChange } from '../common/objects';
```
Could someone elaborate except moving this file and renaming things what was fixes in this file? It is used for all commands triggered by the plugin host process would be bad to break it.

vrubezhny replied:
For `vscode.executeReferenceProvider` and `vscode.executeImplementationProvider`, as they return location array as result,  it is required to convert the result from `model.*` objects to `theia.*` objects in order to be accepted by VSCode Reference View plugin.
So, optional property is added to the mapping to specify the converter function that is to be used to convert the results. As it is optional, there is no need to add this property to all the other existing mappings. Currently, `CONVERT_MONACO_TO_VSCODE` is added to convert the results array into an array of `theia.Location` objects. It's used to convert the results of `vscode.executeReferenceProvider` and `vscode.executeImplementationProvider` command execution.
The only change for the `old` mappings is that `MONACO_CONVERSION_IDENTITY` function used for arguments conversion is renamed to `CONVERT_VSCODE_TO_MONACO` (in order to indicate the fact that it really changes arguments -  requested by @tsmaeder )

tsmaeder replied:
Unfortunately, the github diff algorithm is being really stupid here: there are loads of single line changes in the command definitions, but GH picks it up as a two entirely changed blocks of code 🤷‍♂

akosyakov commented on packages/plugin-ext/src/plugin/tree/tree-views.ts:
```diff
@@ -88,6 +89,22 @@ export class TreeViewsExtImpl implements TreeViewsExt {
             get onDidChangeVisibility() {
                 return treeView.onDidChangeVisibility;
             },
+            // tslint:disable-next-line:typedef
```
or just type string everywhere? it should be even shorter

vrubezhny replied:
I have removed typedefs in favour of specifying type string for getters

akosyakov commented on packages/plugin-ext/src/plugin/tree/tree-views.ts:
```diff
@@ -88,6 +89,22 @@ export class TreeViewsExtImpl implements TreeViewsExt {
+            get message() {
+                return treeView.message;
+            },
+            // tslint:disable-next-line:typedef
```
i don't believe set can typedef, btw we don't use tslint, but eslint

vrubezhny replied:
This was added before we switched to eslint. I'm about to remove these typedef-s, but then should I add return type (string) to be `get message(): string {` and `get title(): string {`? 
Other existing getters do not specify return type...

vrubezhny replied:
(with tslint without typedef added for `set` it was resulting into compilation errors. A bit strange)

akosyakov commented on packages/plugin-ext/src/plugin/tree/tree-views.ts:
```diff
@@ -159,6 +176,10 @@ class TreeViewExtImpl<T> implements Disposable {
     private readonly onDidChangeVisibilityEmitter = new Emitter<TreeViewVisibilityChangeEvent>();
     readonly onDidChangeVisibility = this.onDidChangeVisibilityEmitter.event;
 
+    protected updateMessage = debounce(() => {
```
Why to debounce? Did you observe any performance or usability issues?

vrubezhny replied:
At first invocation of `Find All References` or `Find All Implementations` command, the Reference View widget is not yet created, so I cannot set a message on it. So, when you ask for the references or implementation for the first time the references tree will be shown without that search summary message. Debouncing solves this problem.

tsmaeder replied:
No, we cannot just "do it later". The problem here is that the state of the tree view can be changed before the widget in the UI is actually created. This just means that we need to transfer the state at the time the widget is created. But this can be 5 minutes after creation of the tree view, so the debounce() is not reliable. 
IMO, we have to pass message and title to TreeViewsMain#reveal(). I think that is the only way a plugin can make a tree view visible, right?

tsmaeder replied:
Note sure where to put the message when the widget is not created yet. @akosyakov any suggestions how to do that?

tsmaeder replied:
Had a discussion on Friday and will investigate what the behaviour of `createTreeView` should be: if it is supposed to open the widget, the `set message` should come afterwards and succeed.

vrubezhny replied:
The widget creation happens in several steps and is distributed in several events and backend/frontend communications.
During this process `$setMessage` and `$setTitle` are invoked with default values `message=undefined` and `title="Results"` in order to indicate that the reference tree is empty during the search for the references (the results are not received yet) - trying to process this initial call is the reason of `debounce` appearance.
After the widget creation these methods are called once again with the real (not empty) values.

So, the simple way to get rid of debouncing is just the rejecting to set an empty message (doing nothing when `$setMessage` is invoked with `message=undefined` value). All the following calls to `$setMessage` will not be failed because the widget will be already created

akosyakov replied:
I think it should be part of tree view metadata which initially loaded from package.json. Then `$setMessage` or `$setTitle` are called when metadata should be updated and used during creation. If a view already present when we update it as well.

vrubezhny commented:
>     2. the search term ("CallHierarchyCommands") is  not highlighted
When I test it with Electron, in some cases the result items are highlighted, in other cases are not. In case such an item is not highlighted  it's enough to mouse click on editor area (so the editor receives focus) - the search item immediately gets highlighted.  Exactly the same happens when running Browser version.

>     6. When I do "Show All References" from the context menu in an editor, the references view does not open, I need to go through the menu bar "view" menu.
Yes, it is. In my case I even didn't have `View->References` menu item, it looks like it appeared after I tried to invoke `Find All Referencess` or `Find All Implementations` command. 
But the following invocation of `Find All Referencess`/`Find All Implementations` commands, in my case, successfully  open `References` View (if it's not opened) even after I restarted Electron application several times.

vrubezhny commented:
Updated:
- `$setTitle` is fixed in tree-view-main
- removed tslint disable typedef from settert/getters for `message` and `title` in tree-views

vrubezhny commented:
>     6\. When I do "Show All References" from the context menu in an editor, the references view does not open, I need to go through the menu bar "view" menu.

Fixed for the Electron as well as for the Browser versions.

akosyakov commented on packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts:
```diff
@@ -268,6 +268,24 @@ export class PluginViewRegistry implements FrontendApplicationContribution {
+
+        toDispose.push(this.commands.registerCommand({
+            id: `${view.id}.focus`,
+            label: `Focus ${view.name} View`
```
Are you sure that such commands should have UI contributions. I could not find anything in VS Code, for us it is a bit overkill, since we have other commands. Is it for testing?

vrubezhny replied:
Vscode registeres this command for every registered view:
https://github.com/microsoft/vscode/blob/master/src/vs/workbench/browser/parts/views/views.ts#L542

vscode-references-view plugin uses it to reveal reference view, and probably some other plugins do. So, for sure the plugin needs it and it's not a test command.

akosyakov replied:
I mean if we add `label` then this command is visible in UI like in the command palette. I cannot find such commands visible in VS Code. That we register a command is fine me.

vrubezhny replied:
Yes, no problem. I have removed `label` property. Now the command is not visible and the plugin works fine with this.

akosyakov commented on packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts:
```diff
@@ -268,6 +268,24 @@ export class PluginViewRegistry implements FrontendApplicationContribution {
+            label: `Focus ${view.name} View`
+        }, {
+            execute: async () => {
+                const widget = await this.openView(view.id);
```
I wonder why following does not work:
```
const widget = await this.openView(view.id);
if (widget) {
    this.shell.activateWidget(widget.id);
}
```

It will give the focus to the view, not to containing view container.

vrubezhny replied:
Doing like just `this.shell.activateWidget(widget.id);` doesn't work here.

I had to find view container identifier (because it's registered in shell for the view) and use it to activate the view.

akosyakov replied:
I will need to check, i thought that each view is registered in the shell as well and the shell is capable to handle nested widgets. Focusing view container is not the same as focusing the view.

vrubezhny replied:
Me too, but for that References View there is no reference view`s `widget.id` registered in `shell`'s tracker, 
In `shell` there is an array of tracked widgets that is used to search for a widget to activate by its id in `activateWidget(id: string)`..
So, the id of widget we need to activate is "plugin-view:references-view.tree" while that array contains "plugin-view-container:references-view" for Reference view - that's is why the References view cannot not activated by its view `widgett.id`.

vrubezhny replied:
But anyway, `vscode-references-view` invokes `<view.id>.focus` command in order to activate the Reference View, so if we don't have such a command registered the plugin will not be able to activate the view.

akosyakov replied:
`activateWidget` creates a stack of all tracked widget, so it will be ['plugin-view:references-view.tree', 'plugin-view-container:references-view']. We need to activate all of them. If we activate only the last it will work only till view container has only one view (by luck). For stack it works in following order:
- activate the view container ('plugin-view-container:references-view')
- then ask the view container to activate the view ('plugin-view:references-view.tree')

If it does not work there is a bug in a view that it does not focus properly.

akosyakov replied:
I replaced it with proper activateWidget for recursive activation and could not observer the issue. When does not it work for you? Maybe there is a race between model is rendered and focus is requested?

vrubezhny replied:
Focus command was working with no problem in Browser version but not working in Electron in this version:
```
            execute: async () => {
                const widget = await this.openView(view.id);
                if (widget) {
                    this.shell.activateWidget(widget.id);
                }
            }
```
I'm not sure when, but probably after some rebasing over the master branch, it stopped working, then I found that `plugin-view:references-view.tree` is not in the list of tracked widgets, it contains only `plugin-view-container:references-view` that's why i changed the handler to make activatoin by the view container identifier.

> I replaced it with proper activateWidget for recursive activation and could not observer the issue. When does not it work for you? Maybe there is a race between model is rendered and focus is requested?

This `Focus` command normally is invoked only if the References View is not open, so the handler should open the view and activate it. When I'm debugging it the View widget returned by `const widget = await this.openView(view.id);` is valid (not `null` or `undefined`), but its `id` is not on the list of tracked widgets, so the activation didn't worked.
If you tell that the widget `id` must be added to the list of tracked widgets, then probably we should look why it is not added there. (Without having it there even recursive activation probably will still be not full and correct?)

> I replaced it with proper activateWidget for recursive activation...

Did you made it locally in your bramch? Where can I look at it?

vrubezhny commented:
PR is updated:
- no more rebounce, the `$setMessage` can be invoked at any time whether the widget is created on Tree Vi ew or it's not
- The css style is applied on the message for its indentation 
- Command execution results now  may be any value, array or undefined

akosyakov commented:
Do I get correctly that the reference widget extension already makes use of the call hierarchy API? And we can replace our call hierarchy Theia extension by using it?

vrubezhny commented:
> Do I get correctly that the reference widget extension already makes use of the call hierarchy API? And we can replace our call hierarchy Theia extension by using it?

No. 
Yes it uses CallHierarchy API, but I really never could make it work ]from the commands provided by the current version of the extension] on VSCode when I was testing it. The plugin's REAAD.me doesn't have any information on CallHierarchy commands... So, I suppose it either just doesn't work with typescript or there are some other limitations.

The support for `vscode.prepareCallHierarchy`, `vscode.provideIncomingCalls` and `vscode.provideOutgoingCalls` commands is included in order to fully support vscode-reference-view

vrubezhny commented:
The `<view.id>.focus` command registration is made to not have UI contributions.

akosyakov commented:
I have errors in logs when try to activate it with keybindings:
<img width="1356" alt="Screenshot 2020-02-21 at 16 22 47" src="https://user-images.githubusercontent.com/3082655/75047041-7025ba80-54c6-11ea-9962-fd17242362f9.png">

akosyakov commented:
There is no enough spacing around message:
<img width="504" alt="Screenshot 2020-02-21 at 16 23 10" src="https://user-images.githubusercontent.com/3082655/75047102-8764a800-54c6-11ea-9877-d0818645ac52.png">

And in VS Code:
<img width="409" alt="Screenshot 2020-02-21 at 16 23 16" src="https://user-images.githubusercontent.com/3082655/75047094-8469b780-54c6-11ea-80de-141f1363595e.png">

akosyakov commented on packages/plugin-ext/src/main/browser/view/plugin-view-widget.ts:
```diff
@@ -59,6 +60,7 @@ export class PluginViewWidget extends Panel implements StatefulWidget {
         const widget = this.widgets[0];
         if (widget) {
             widget.activate();
+            this.updateWidgetMessage();
```
it should not be here, activate means focus something in already rendered widget

vrubezhny replied:
`this.updateWidgetMessage` has nothing to do with `PluginViewWidget`'s rendfering. It just  "pushes" the 'message'  value further into the `TreeViewWidget` in case it is already created and added here.
The way `onActivateRequest` method is written:
```
        const widget = this.widgets[0];
        if (widget) {
            widget.activate();
        } else {
            this.node.focus();
        }
```
tells nothing about that the widget is already rendered.  Imho, it's not even initialized properly in case of `this.widgets[0]; === undefined`, is it?

akosyakov commented on packages/plugin-ext/src/main/browser/view/plugin-view-widget.ts:
```diff
@@ -96,10 +100,39 @@ export class PluginViewWidget extends Panel implements StatefulWidget {
+        }
+    }
+
+    addWidget(widget: Widget): void {
```
we should revert `addWidget` and `insertWidget` they are not for rendering.

vrubezhny replied:
It is not about rendering. `addWidget` and `insertWidget` are overridden in order to safely set message value on the widget, because `$setMessage` API is invoked mostly when the widget creation is started, but not finished (so, the widget is not set into the View's widgets[] and, as such is not accessible).  
Otherwise, I don't know a way to properly realize `$setMessage` API other than `debouncing` which we were trying to get rid of. Any suggestions?

akosyakov replied:
Let's stick with it for now and wait for a bug. Such approach seems to work timely wise concretely for this extension. If all extension reveal the widget before setting message then it should work generally.

akosyakov commented on packages/plugin-ext/src/main/browser/view/plugin-view-widget.ts:
```diff
@@ -96,10 +100,39 @@ export class PluginViewWidget extends Panel implements StatefulWidget {
+    private updateWidgetMessage(): void {
+        const widget = this.widgets[0];
+        if (widget) {
+            if (widget instanceof TreeViewWidget) {
```
we should set `message` directly to `TreeViewWidget`, `PluginVIewWidget` does not need to deal with it

vrubezhny replied:
Again, `TreeViewWidget` is not accessible at the moment of `$setMessage` call, this is why I save the message in `PluginVIewWidget` untill `addWidget` or `insertWidget` is invoked.

akosyakov commented on packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx:
```diff
@@ -372,4 +372,25 @@ export class TreeViewWidget extends TreeWidget {
+    }
+
+    set message(message: string | undefined) {
+        this._message = message;
```
pleas guard if message does not change to avoid unnecessary rendering

vrubezhny replied:
OK.

akosyakov commented on packages/plugin-ext/src/main/browser/view/plugin-view-widget.ts:
```diff
@@ -67,12 +69,14 @@ export class PluginViewWidget extends Panel implements StatefulWidget {
     storeState(): PluginViewWidget.State {
         return {
             label: this.title.label,
+            message: this.message,
```
it has to be a part of `TreeViewWidget` state

vrubezhny replied:
I can do so, but the question is the same - how to access it during the tree view creation, when `TreeViewWidget` is not added to `PluginViewWidget` yet?

vrubezhny commented:
> There is no enough spacing around message and the color is bogus:
> <img alt="Screenshot 2020-02-21 at 16 23 10" width="504" src="https://user-images.githubusercontent.com/3082655/75047102-8764a800-54c6-11ea-9877-d0818645ac52.png">
> 
> And in VS Code:
> <img alt="Screenshot 2020-02-21 at 16 23 16" width="409" src="https://user-images.githubusercontent.com/3082655/75047094-8469b780-54c6-11ea-80de-141f1363595e.png">
> 

I'll fix spacing

> BTW what is wrong with icons, folders instead of files?
> 

Probably we should look into Label Provider: https://github.com/eclipse-theia/theia/blob/master/packages/core/src/browser/label-provider.ts#L203
I'd like to create a separate issue for this.

> Not sure why description color is wrong as well. We need to make it look decent at least and then file issues for everything else if you don't want to address them in this PR.

What do you mean by "description color"? Is it a `X results in Y files` (message) on top of the tree?
As for me, the View colors in Chrome (for Browser example) and in VSCode are pretty the same (Fedora 29, Gnome):
![Screenshot from 2020-02-21 22-46-51](https://user-images.githubusercontent.com/620781/75074796-b990fc80-54fc-11ea-823d-2d977bc1b447.png)
The only that `message` is different (grey instead of white). 
I don't have `orange` backgrounded words (looks like a word that was searched for) - but, imho, it's better not to have it colored on the tree side.
But I don't understand why some lines in your VSCode have `orange` foreground and a labeled like `9+` on the right... Is it a something in the UI specially made for MacOS?

vrubezhny commented:
> I have errors in logs when try to activate it with keybindings:
> <img alt="Screenshot 2020-02-21 at 16 22 47" width="1356" src="https://user-images.githubusercontent.com/3082655/75047041-7025ba80-54c6-11ea-9962-fd17242362f9.png">

I don't have any errors in log when I use `Alt + Shift + F12` to find references. The hotkey combination is not shown in context menu, but it still works fine for me.
What version you use to test -  Browser or Electron?

akosyakov commented:
@vrubezhny Could you please do https://github.com/eclipse-theia/theia/pull/7153#issuecomment-589696647 it will easy testing of this PR and it would be good to have this extension on master after merging.

vrubezhny commented:
I've updated the PR with:
- Margins and foregroung color are fixed for the Tree View Message
- PR #7153 is merged into this one (adds vscode-references-view plugin to the list of plugins to be downloaded from https://download.jboss.org at Theia build, so making it available for use in run time)

akosyakov commented on package.json:
```diff
@@ -142,6 +142,7 @@
     "vscode-editorconfig": "https://github.com/theia-ide/editorconfig-vscode/releases/download/v0.14.4/EditorConfig-0.14.4.vsix",
-    "vscode-eslint": "https://github.com/theia-ide/vscode-eslint/releases/download/release%2F2.0.15/vscode-eslint-2.0.15.vsix"
+    "vscode-eslint": "https://github.com/theia-ide/vscode-eslint/releases/download/release%2F2.0.15/vscode-eslint-2.0.15.vsix",
+    "vscode-references-view": "https://download.jboss.org/jbosstools/vscode/3rdparty/vscode-references-view/references-view-0.0.47-05339d.vsix"
```
please use this link https://github.com/theia-ide/vscode-references-view/releases/download/v0.0.47/references-view-0.0.47.vsix

For now we fork all extension under thiea-ide org, later we will consume from our registry. That's how it should be done, not pulled from somewhere else.

vrubezhny replied:
Done. Sorry, I didn't spot the change in URL in your comment to #7153

akosyakov commented:
> https://github.com/eclipse-theia/theia/pull/7055#issuecomment-589863848

@vrubezhny I can reproduce it reliably for browser. Please make sure to reset the layout before that it is not opened.

akosyakov commented on packages/plugin-ext/src/main/browser/style/tree.css:
```diff
@@ -38,3 +38,10 @@
 }
+
+.theia-tree-view  .theia-TreeContainer .theia-TreeViewInfo {
+    color: var(--theia-list-inactiveSelectionForeground);
```
i don't think you need to declare it at all, it will inherit proper foreground

akosyakov replied:
@vrubezhny Could you remove this color please after that I think we should merge it?
@tsmaeder would it be fine with you or you want to have a look again?

vrubezhny replied:
Yes, now it works. However, previously a `gray`-like color was selected as the foreground without this definition.

akosyakov commented:
It looks good except keybindings don't work for me.

akosyakov commented:
@vrubezhny Don't care about commands. I know there is the issue and will be fixed in: https://github.com/eclipse-theia/theia/pull/6921/commits/dae68659bcd861fce2d395e11ebe5028d985c23f#diff-0291e1b92ff01c13c0dc1c98263ac426 Keybindings don't call command handlers properly, so the plugin system gets never notified about executed commands.
